explainers.md

```diff
@@ -0,0 +1,860 @@
+
+On desktop:
+
+1. Apps without an ID field continue to use start\_url to generate ids.
+2. New apps with an ID field specified will use the new ID.
+3. Apps that already exist that want to use the ID field will specify both an ID field and **legacy_ids_** field. This instructs user agents to look up installed apps by legacy_ids to do a **backend** **migration** to a new ID scheme.
+
+Alternatively, we can have separate legacy_manifest_url and legacy_start_url field to be explicit. The Cons with this approach is we have one more field and they are platform-specific.
+
+Pros:
+
+*   Be specific so the app developers know exactly which field they are migrating from.
+
+Cons:
+
+*   Platform specific and not generic. This could be a anti-pattern for manifest design.
+
+When legacy_ids is presented, exactly how to migrate will be an implementation detail of the user agent.
+
+For Chromium desktop, these are the possible ways to do the migrate:
+
+Plan A:
+
+When a manifest is loaded, look up existing apps with id = &lt;new id>/&lt;legacy_ids>. 
+
+If found, mark the old app as “deprecated”. Hide it from chrome://apps. create a new app. Remove local shortcuts for the old app.
+
+If not found, save the new app as the new id.
+
+Plan B:
+
+When a manifest is loaded, look up existing apps with id = &lt;new id> or &lt;legacy_ids>. 
+
+If found, apply updates without changing the id.
+
+If not found, save the new app as the new id.
+
+Plan A could cause some friction when swapping the local shortcuts and app instances. Any local changes saved for the old app before the user closes the old app won’t be transferred to the new local app.
+
+Plan B won’t have this problem, but it will leave installed apps with old ids for a longer time.
+
+For **previous versions of Chrome**, if it receives a sync entity from other devices with the new ID, it will be recognized as a separate app. 
+
+
+## Conclusion
+
+Using existing manifest\_url/start\_url would be bad because there are multiple use cases that require changing them. There are ways to mitigate changing of manifest\_urls if old ones can still be supported, but they are not ideal and don’t satisfy all the developers' needs. Having the global\_id to be resolvable is a pros for using manifest\_url, but it’s a “nice to have”, but not a required feature for the id. The main goal for the unique id is still to ensure having a stable identifier for apps that are flexible with updating their other metadata.
+
+Using **start\_url\_origin + id** would be preferred as this is the most stable option.
+
+For the default field to use, **start\_url** is preferred.
+
+Because of the existing sync design of desktop PWAs, old chromes will be unavoidable to have duplicate apps if the default is not start\_url, there might be weird behaviors when duplicate apps exist.
+
+Due to the use cases for data URL, cross origin manifest URL, as well as the sync behavior, using start\_url is preferred over manfiest\_url as default.
+
+
```

Comment:
Might be nice to acknowledge the discoverability issue, and welcome ideas to address this.

Reply:
Mentioned this in the beginning of the conclusion